src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// to be recursively batched in the same ep_rt_mono_log_type_and_parameters call
+	if ((mono_underlying_type == MONO_TYPE_ARRAY) || (mono_underlying_type == MONO_TYPE_SZARRAY))
+	{
+		MonoArrayType *mono_array_type = mono_type_get_array_type (mono_type);
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_ARRAY;
+		if (mono_underlying_type == MONO_TYPE_ARRAY)
+		{
+			// Only ranks less than kEtwTypeFlagsArrayRankMax are supported.
+			// Fortunately kEtwTypeFlagsArrayRankMax should be greater than the
+			// number of ranks the type loader will support
+			uint32_t rank = mono_array_type->rank;
+			if (rank < K_ETW_TYPE_FLAGS_ARRAY_RANK_MAX)
+			{
+				rank <<= 8;
+				p_val->fixed_sized_data.flags |= rank;
+			}
+		}
+
+		// mono arrays are always arrays of by value types
+		p_val->rg_mono_type_parameters[p_val->c_type_parameters] = m_class_get_byval_arg (mono_array_type->eklass);
+		p_val->rg_type_parameters[p_val->c_type_parameters] = get_typeid_for_type (m_class_get_byval_arg (mono_array_type->eklass));
+		p_val->c_type_parameters++;
+	}
+	else if (mono_underlying_type == MONO_TYPE_GENERICINST)
+	{
+		MonoGenericInst *class_inst = mono_type->data.generic_class->context.class_inst;
+		p_val->c_type_parameters = class_inst->type_argc;
+		for (int i = 0; i < class_inst->type_argc; i++)
+		{
+			p_val->rg_mono_type_parameters[i] = class_inst->type_argv[i];
+			p_val->rg_type_parameters[i] = get_typeid_for_type (class_inst->type_argv[i]);
+		}
+	}
+	else if (mono_underlying_type == MONO_TYPE_CLASS || mono_underlying_type == MONO_TYPE_VALUETYPE || mono_underlying_type == MONO_TYPE_PTR || mono_underlying_type == MONO_TYPE_BYREF)
+	{
+		p_val->rg_mono_type_parameters[p_val->c_type_parameters] = mono_type_get_underlying_type (mono_type);
+		p_val->rg_type_parameters[p_val->c_type_parameters] = get_typeid_for_type (mono_type_get_underlying_type (mono_type));
+		p_val->c_type_parameters++;
+	}
+
+	if (mono_class_has_finalizer (klass))
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_FINALIZABLE;
+
+	if (m_class_is_delegate (klass))
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_DELEGATE;
+
+	if (mono_class_is_com_object (klass))
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_EXTERNALLY_IMPLEMENTED_COM_OBJECT;
+
+	// Now that we know the full size of this type's data, see if it fits in our
+	// batch or whether we need to flush
+	int cb_val = ep_rt_mono_get_byte_count_in_event(p_val);
+	if (cb_val > k_max_bytes_type_values)
+	{
+		if (p_val->s_name)
+			p_val->s_name[0] = '\0';
+		cb_val = ep_rt_mono_get_byte_count_in_event(p_val);
+
+		if (cb_val > k_max_bytes_type_values)
+		{
+			// This type is apparently so huge, it's too big to squeeze into an event, even
+			// if it were the only type batched in the whole event.  Bail
+			return -1;
+		}
+	}
+	if (p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)
+	{
+		// Although this type fits into the array, its size is so big that the entire
```

Comment:
In addition to this comment, can we put an assert or other debug time mechanisms for validating this assertion?

Reply:
What exactly are we asserting? Just that the if statement holds? `EP_ASSERT(p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)`? Is this so that in debug mode we know that this was hit?

Reply:
The statement in the assert expression needs to always be true, or you will get an assert fired that will break the process with a message (the failing expression). EP_ASSERT is mapped to a g_assert_checked on Mono so it's only in checked build.

Reply:
Added a g_assert, but would it be more helpful to have an assert for the opposite `p_type_logger->m_n_bulk_type_value_byte_count + cb_val <= k_max_bytes_type_values` after this if?